### PR TITLE
Fixes for some object ownership issues with applications, groups and service principals

### PR DESCRIPTION
**Pull request by manicminer**
This PR is functionally complete but pending https://github.com/manicminer/hamilton/pull/86

_For applications and service principals:_

Where no owners are specified in configuration, The calling principal is explicitly assigned as the owner at creation time, and then removed immediately afterwards. This provides stable behavior across resources where sometimes the API will assign the caller as the owner and sometimes not, and it's impossible to explicitly specify zero owners at creation time.

Where owners are specified, 19 of them are assigned at creation time, in addition to the calling principal. This is unordered since the `owners` property is a TypeSet in all cases. The remaining owners (if any) are subsequently assigned immediately after creation. If the calling principal isn't included in the configured owners, it's removed immediately afterwards.

_For groups:_

**API Issues**

- New groups are assigned the calling principal as the sole owner ([contrary to documentation](https://docs.microsoft.com/en-us/graph/api/group-post-groups?view=graph-rest-beta))
- It's impossible to create a group with no owners ([contrary to documentation](https://docs.microsoft.com/en-us/graph/api/group-post-groups?view=graph-rest-beta))
  - Empty list: `The value of 'odata.bind' property annotation is an empty array. The 'odata.bind' property annotation must have a non-empty array as its value`
  - null: `Invalid value for owners@odata.bind property`
- Groups cannot have all their owners removed ([documented](https://docs.microsoft.com/en-us/graph/api/group-delete-owners?view=graph-rest-beta))
- M365 groups must have at least one owner that is a user ([undocumented](https://docs.microsoft.com/en-us/microsoft-365/admin/create-groups/office-365-groups))
  - This is inconsistently enforced. It's possible to create an M365 group with a single service principal owner at the expense of further validations errors when later updating owners.
- Security / legacy groups have erroneously inherited some of the validation for M365 group user owners
  - When removing the last _user_ owner from a security group, this error is returned: `The group must have at least one owner, hence this owner cannot be removed.`

**Remediation**

If no owners are specified in configuration, the calling principal is assigned ownership and retains that ownership. The `owners` property is optional + computed.

Where owners are specified, the calling principal is _not_ given ownership. The first (up to) 20 principals specified in configuration are preferentially selected to be the calling principal (if included in the configuration), then any users specified, followed by any other service principals - these are included in the POST request to create the group. Any additional owners are added one-by-one post-creation of the group.

This aims to work around strict validation with Microsoft 365 groups where at least one owner must be a user at all times! This validation has also been observed erroneously with legacy / security groups too. The API also prevents removal of the last remaining user owner, if a user has ever been assigned ownership of a group. A group cannot have all its owners removed (hence `MinItems: 1`). Ensuring that at least one user is included in the set of owners in configuration is an exercise left to the practitioner.

Unfortunately very little of this can be validated at plan time, since the owner principals may or may not exist at that time. Have opted to leave the API errors around owner removal exposed at apply time, since any validation we add at this point is papering over that and doesn't improve things.

Create and Update timeouts are also increased for groups to 20 mins, and applications & service princpals to 10 mins, to accommodate the time needed to post large numbers of members/owners.

Closes: #478 
Closes: #517

## Enhancements

- `azuread_service_principal` - add support for the `owners` property

### CODE REVIEW

**Review by katbyte — APPROVED** [review]
Thanks, LGTM aside from some comments on var naming - i've only commented in one place but it applies to all the places following that pattern.

File: internal/services/applications/application_resource.go [hunk 1]
```diff
@@ -846,6 +850,52 @@ func applicationResourceCreate(ctx context.Context, d *schema.ResourceData, meta
...
+	// Track whether we need to remove the calling principal later on
+	removeCallerOwner := true
+
+	// Retrieve and set the initial owners, which can be up to 20 in total when creating the application
+	if v, ok := d.GetOk("owners"); ok {
+		c := 0
+		for _, id := range v.(*schema.Set).List() {
+			i := 0
+			if c >= 19 {
+				i = 1
```
[the remark this answers is not in the document]
Comment (manicminer) [reply internal/services/applications/application_resource.go]: Good call, counter variable changed to `ownerCount`

File: internal/services/groups/group_resource.go [hunk 2]
```diff
@@ -408,8 +401,88 @@ func groupResourceCreate(ctx context.Context, d *schema.ResourceData, meta inter
 		properties.Visibility = utils.String(visibility)
 	}
 
-	// The calling principal is the initial owner (retained after other owners are also set)
-	properties.AppendOwner(client.BaseClient.Endpoint, client.BaseClient.ApiVersion, callerId)
+	// Chunk the owners into two slices, the first containing up to 20 and the rest overflowing to the second slice
+	ownerChunks := make([]msgraph.Owners, 2)
```
[the remark this answers is not in the document]
Comment (manicminer) [reply internal/services/groups/group_resource.go]: Makes sense, have changed this to use two variables instead of a slice of slices 👍

**Comment by manicminer** [comment]
### Test results

#### Applications

Screenshot 2021-08-20 at 00 52 00

#### Groups

One unrelated failure, one failure attributable to as-yet-unhanded replication delay with groups having dozens of owners (users unlikely to observe, to be fixed separately)

Screenshot 2021-08-20 at 00 52 18

#### Service Principals

One failure attributable to as-yet-unhanded replication delay with service principals having dozens of owners (users unlikely to observe, to be fixed separately)

Screenshot 2021-08-20 at 00 52 34

#### Users

Screenshot 2021-08-20 at 01 18 02

**Pull request merged**